Arc - Method used by Yasson not implemented

Adding a @JsonbTypeSerializer or other annotations that lookup for a Bean, this Exception is thrown:

```
Caused by: java.lang.UnsupportedOperationException
	at io.quarkus.arc.BeanManagerImpl.createAnnotatedType(BeanManagerImpl.java:211)
	at org.eclipse.yasson.internal.components.BeanManagerInstanceCreator.lambda$getOrCreateComponent$0(BeanManagerInstanceCreator.java:68)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
	at org.eclipse.yasson.internal.components.BeanManagerInstanceCreator.getOrCreateComponent(BeanManagerInstanceCreator.java:67)
	at org.eclipse.yasson.internal.ComponentMatcher.introspectSerializerBinding(ComponentMatcher.java:292)
	at org.eclipse.yasson.internal.AnnotationIntrospector.getSerializerBinding(AnnotationIntrospector.java:299)
	at org.eclipse.yasson.internal.AnnotationIntrospector.introspectCustomization(AnnotationIntrospector.java:694)
	at org.eclipse.yasson.internal.MappingContext$ParseClassModelFunction.apply(MappingContext.java:54)
	at org.eclipse.yasson.internal.MappingContext$ParseClassModelFunction.apply(MappingContext.java:37)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
	at org.eclipse.yasson.internal.MappingContext.getOrCreateClassModel(MappingContext.java:107)
	at org.eclipse.yasson.internal.Marshaller.marshall(Marshaller.java:73)
	at org.eclipse.yasson.internal.JsonBinding.toJson(JsonBinding.java:98)
	at org.jboss.resteasy.plugins.providers.jsonb.JsonBindingProvider.writeTo(JsonBindingProvider.java:133)
	... 60 more
```

Is documented that createAnnotatedType is not supported but JSONB RI uses it as well (I think is older than Quarkus itself).

--- a/bom/pom.xml
+++ b/bom/pom.xml
@@ -157,7 +157,7 @@
         <jboss-logmanager.version>1.0.3</jboss-logmanager.version>
         <jetty.version>9.4.18.v20190429</jetty.version>
         <flyway.version>5.2.4</flyway.version>
-        <yasson.version>1.0.3</yasson.version>
+        <yasson.version>1.0.4</yasson.version>
         <!-- Used for integration tests, to make sure webjars work-->
         <bootstrap.version>3.1.0</bootstrap.version>
         <keycloak.version>6.0.1</keycloak.version>

--- a/extensions/camel/camel-servlet/runtime/src/main/java/io/quarkus/camel/servlet/runtime/CamelServletConfig.java
+++ b/extensions/camel/camel-servlet/runtime/src/main/java/io/quarkus/camel/servlet/runtime/CamelServletConfig.java
@@ -3,8 +3,6 @@ package io.quarkus.camel.servlet.runtime;
 import java.util.List;
 import java.util.Map;
 
-import javax.servlet.Servlet;
-
 import io.quarkus.runtime.annotations.ConfigGroup;
 import io.quarkus.runtime.annotations.ConfigItem;
 import io.quarkus.runtime.annotations.ConfigPhase;

--- a/extensions/jsonb/deployment/pom.xml
+++ b/extensions/jsonb/deployment/pom.xml
@@ -20,6 +20,10 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-arc-deployment</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-jsonp-deployment</artifactId>
         </dependency>
         <dependency>

--- a/extensions/jsonb/deployment/src/main/java/io/quarkus/jsonb/deployment/JsonbProcessor.java
+++ b/extensions/jsonb/deployment/src/main/java/io/quarkus/jsonb/deployment/JsonbProcessor.java
@@ -15,21 +15,51 @@
  */
 package io.quarkus.jsonb.deployment;
 
-import org.eclipse.yasson.JsonBindingProvider;
+import java.util.function.Predicate;
 
+import javax.json.bind.adapter.JsonbAdapter;
+
+import org.eclipse.yasson.JsonBindingProvider;
+import org.eclipse.yasson.spi.JsonbComponentInstanceCreator;
+import org.jboss.jandex.DotName;
+
+import io.quarkus.arc.deployment.BeanArchiveIndexBuildItem;
+import io.quarkus.arc.deployment.UnremovableBeanBuildItem;
+import io.quarkus.arc.processor.BeanInfo;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.substrate.ReflectiveClassBuildItem;
+import io.quarkus.deployment.builditem.substrate.ServiceProviderBuildItem;
 import io.quarkus.deployment.builditem.substrate.SubstrateResourceBundleBuildItem;
+import io.quarkus.jsonb.QuarkusJsonbComponentInstanceCreator;
 
 public class JsonbProcessor {
 
+    static final DotName JSONB_ADAPTER_NAME = DotName.createSimple(JsonbAdapter.class.getName());
+
     @BuildStep
     void build(BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
-            BuildProducer<SubstrateResourceBundleBuildItem> resourceBundle) {
+            BuildProducer<SubstrateResourceBundleBuildItem> resourceBundle,
+            BuildProducer<ServiceProviderBuildItem> serviceProvider) {
         reflectiveClass.produce(new ReflectiveClassBuildItem(false, false,
                 JsonBindingProvider.class.getName()));
 
         resourceBundle.produce(new SubstrateResourceBundleBuildItem("yasson-messages"));
+
+        serviceProvider.produce(new ServiceProviderBuildItem(JsonbComponentInstanceCreator.class.getName(),
+                QuarkusJsonbComponentInstanceCreator.class.getName()));
     }
+
+    @BuildStep
+    void unremovableJsonbAdapters(BuildProducer<UnremovableBeanBuildItem> unremovableBeans,
+            BeanArchiveIndexBuildItem beanArchiveIndex) {
+        unremovableBeans.produce(new UnremovableBeanBuildItem(new Predicate<BeanInfo>() {
+
+            @Override
+            public boolean test(BeanInfo bean) {
+                return bean.isClassBean() && bean.hasType(JSONB_ADAPTER_NAME);
+            }
+        }));
+    }
+
 }

--- a/extensions/jsonb/runtime/pom.xml
+++ b/extensions/jsonb/runtime/pom.xml
@@ -28,6 +28,10 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-arc</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-jsonp</artifactId>
         </dependency>
     </dependencies>

--- a/extensions/jsonb/runtime/src/main/java/io/quarkus/jsonb/QuarkusJsonbComponentInstanceCreator.java
+++ b/extensions/jsonb/runtime/src/main/java/io/quarkus/jsonb/QuarkusJsonbComponentInstanceCreator.java
@@ -1,0 +1,50 @@
+package io.quarkus.jsonb;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.eclipse.yasson.spi.JsonbComponentInstanceCreator;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.InstanceHandle;
+
+public class QuarkusJsonbComponentInstanceCreator implements JsonbComponentInstanceCreator {
+
+    private final Map<Class<?>, Object> components;
+
+    // Instance handles are used to destroy the CDI-based components correctly
+    private final List<InstanceHandle<?>> beanHandles;
+
+    public QuarkusJsonbComponentInstanceCreator() {
+        this.beanHandles = new ArrayList<>();
+        this.components = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public void close() throws IOException {
+        beanHandles.forEach(InstanceHandle::close);
+        beanHandles.clear();
+        components.clear();
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> T getOrCreateComponent(Class<T> componentClass) {
+        return (T) components.computeIfAbsent(componentClass, c -> {
+            InstanceHandle<T> beanHandle = Arc.container().instance(componentClass);
+            if (beanHandle.isAvailable()) {
+                beanHandles.add(beanHandle);
+                return beanHandle.get();
+            }
+            try {
+                return componentClass.newInstance();
+            } catch (InstantiationException | IllegalAccessException e) {
+                throw new IllegalStateException("Cannot instantiate Jsonb component: " + componentClass, e);
+            }
+        });
+    }
+
+}

--- a/extensions/jsonb/runtime/src/main/resources/META-INF/services/org.eclipse.yasson.spi.JsonbComponentInstanceCreator
+++ b/extensions/jsonb/runtime/src/main/resources/META-INF/services/org.eclipse.yasson.spi.JsonbComponentInstanceCreator
@@ -1,0 +1,1 @@
+io.quarkus.jsonb.QuarkusJsonbComponentInstanceCreator

--- a/independent-projects/arc/pom.xml
+++ b/independent-projects/arc/pom.xml
@@ -51,7 +51,7 @@
         <maven.compiler.target>1.8</maven.compiler.target>
         <!-- Versions -->
         <version.cdi>2.0.SP1</version.cdi>
-        <version.jandex>2.1.0.Final</version.jandex>
+        <version.jandex>2.1.1.Final</version.jandex>
         <version.junit4>4.12</version.junit4>
         <version.maven>3.5.2</version.maven>
         <version.jboss-logging>3.3.2.Final</version.jboss-logging>

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java
@@ -213,6 +213,15 @@ public class BeanInfo implements InjectionTargetInfo {
         return types;
     }
 
+    public boolean hasType(DotName typeName) {
+        for (Type type : types) {
+            if (type.name().equals(typeName)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
     public Set<AnnotationInstance> getQualifiers() {
         return qualifiers;
     }

